src/plugins/intel_gpu/src/plugin/infer_request.cpp

```diff
@@ -813,14 +824,53 @@ std::map<std::string, InferenceEngineProfileInfo> InferRequest::GetPerformanceCo
     }
 }
 
+void InferRequest::allocate_dev_mem_if_needed(InferenceEngine::BlobMap& device_mems, InferenceEngine::Blob::Ptr& user_blob,
+                                              const cldnn::primitive_id& blob_name, const cldnn::layout& layout) {
+    const auto input_ptr = static_cast<const void*>(user_blob->cbuffer());
+    const auto alloc_type = m_graph->GetEngine()->detect_allocation_type(input_ptr);
+    const auto is_usm_host = alloc_type == cldnn::allocation_type::usm_host;
+    const auto has_device_blob = device_mems.find(blob_name) != device_mems.end();
+    bool can_skip_allocation = false;
+
+    if (has_device_blob) {
+        auto impl = getBlobImpl(device_mems[blob_name]->as<gpu::ClBlob>());
+
+        OPENVINO_ASSERT(impl, str_device_output_unsupported_blob);
+        OPENVINO_ASSERT(impl->is_allocated(), str_input_not_allocated);
+
+        auto impl_mem = impl->getMemory();
+        auto src_ptr = user_blob->cbuffer().as<uint8_t*>();
+        // If device mem already exists, we can reuse blob if buffer has usm_host type and points to the same memory,
+        // so we don't need to allocate new memory
+        can_skip_allocation |= same_host_mem(impl_mem, src_ptr);
+        // Or if blob has any type except usm_host - in that case explicit copy will be performed anyway
+        can_skip_allocation |= impl_mem->get_allocation_type() != cldnn::allocation_type::usm_host;
+    }
+
+    if (!can_skip_allocation) {
+        if (is_usm_host) {
+            // For USM case we create host blob using custom USM host allocator
+            // and then create shared device blob on top of this buffer
+            device_mems[blob_name] = create_shared_device_blob(user_blob->getTensorDesc(), layout, user_blob->buffer().as<void*>());
```

Comment:
may I know if any extra overhead if need to create device blob frequently when user_blob changes from time to time?